<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследуем-данные-и-ответим-на-вопросы" data-toc-modified-id="Исследуем-данные-и-ответим-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуем данные и ответим на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надежности заемщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель проекта:** На основе статистики о платёжеспособности клиентов исследовать влияет ли семейное положение и количество детей клиента на факт возврата кредита в срок.

**Этапы исследования:**
1. Загрузка и обзор данных
2. Предобработка данных (удаление пропусков, обработка аномальных значений, изменение типов данных, обработка дубликатов, категоризация данных)
3. Анализ зависимости между фактом возврата кредита в срок и такими параметрами как количество детей, семейное положение, уровень дохода, цель кредита.

**Описание данных**
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

## Загрузка и обзор данных

Импортируем необходимые библиотеки.  

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats

Считаем данные из csv-файла в датафрейм и сохраните в переменную `data`.

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

Выведем первые 20 строк датафрейма `data` на экран.

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Выведем основную информацию о датафрейме с помощью метода `info()`.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод по шагу 1**

Данные загружены. Осуществлен первичный обзор данных.

## Предобработка данных

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца, используя комбинацию  методов `isna()` и `sum()`.

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [6]:
for t in data['income_type'].unique(): 
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработаем значения в этом столбце, заменив все отрицательные значения положительными с помощью метода `abs()`.

In [7]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.

In [8]:
data.groupby('income_type')['days_employed'].median()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Этот столбец не понадобится для исследования.

Выведем перечень уникальных значений столбца `children`.

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце `days_employed` медианными значениями для каждого типа занятости `income_type`.

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены. Ещё раз выведем количество пропущенных значений для каждого столбца с помощью методов `isna()` и `sum()`.

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`

In [14]:
data['total_income'] = data['total_income'].astype('int')

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [15]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. При наличии таких строк удалим их.

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.

In [18]:
def categorize_income(income): 
    if income <= 30000:
        return 'E'
    if income >= 30001 and income <= 50000:
        return 'D'
    if income >= 50001 and income <= 200000:
        return 'C'
    if income >= 200001 and income <= 1000000:
        return 'B'
    if income >= 1000001:
        return 'A'

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран первые строки датасета для проверки.

In [20]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C

Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.

Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут  правильно определить категорию.

In [22]:
def categorize_purpose(purpose):
    if 'автомоб' in purpose:
        return 'операции с автомобилем' 
    if 'недвиж' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свад' in purpose:
        return 'проведение свадьбы'
    if 'образов' in purpose:
        return 'получение образования'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

Выведем первые 5 строк датасета, чтобы посмотреть результат.

In [24]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

**Выводы по шагу 2**

Осуществлена предобработка данных:
* заполнены пропуски: пропуски в столбцах `total_income` и `days_employed` заполнены медианными значениями по каждому типу занятости из столбца `income_type`;
* обработаны аномальные значения: отрицательные значения в столбце `days_employed` преобразованы в положительные, строки с аномальными значениями в столбце `children` удалены;
* тип данных в столбце `total_income` изменен на целочисленный;
* созданы новые столбцы `total_income_category` и `purpose_category` с категориальными данными о доходе и целях кредита.

## Исследуем данные и ответим на вопросы

Вызовем актуальную информацию о данных в таблице `data`. Теперь в таблице 21331 строка. Пропущенных значений нет.

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int64  
 1   days_employed          21331 non-null  float64
 2   dob_years              21331 non-null  int64  
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int64  
 10  total_income           21331 non-null  int64  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(1), int64(6), object(7)
memory usage: 2.4+ 

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Построим сводную таблицу, показывающую соотношение случаев возврата и невозврата в срок кредита по группам клиентов с разным количеством детей. 
Как было установлено ранее, столбец `'children'` содержит значения от 0 (нет детей) до 5. Значения в столбце `'debt'`: 1 - наличие задолженности, 0 - отсутствие задолженности. 
В качестве значения аргумента `values` можно принять любой столбец (кроме `'children'` и `'debt'`).
В качестве значения аргумента `aggfunc` принимаем `'count'` - подсчитываем количество строк с каждым из значений в столбце `'debt'`.

In [26]:
children_debt_pivot = data.pivot_table(index=['children'], columns='debt', values='education', aggfunc='count')
children_debt_pivot

debt            0       1
children                 
0         13028.0  1063.0
1          4364.0   444.0
2          1858.0   194.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

Заменим пропуск (NaN) на 0

In [27]:
children_debt_pivot =  children_debt_pivot.fillna(0)
children_debt_pivot

debt            0       1
children                 
0         13028.0  1063.0
1          4364.0   444.0
2          1858.0   194.0
3           303.0    27.0
4            37.0     4.0
5             9.0     0.0

Убедимся, что сумма значений в `children_debt_pivot` совпадает с количеством строк таблицы `data` - 21331.

In [28]:
sum(children_debt_pivot[0])+sum(children_debt_pivot[1])

21331.0

Добавим в таблицу `children_debt_pivot` столбец `'debt_percentage'`, показывающий долю случаев несвоевременного возврата кредита в каждой группе клиентов по количеству детей, и отсортируем таблицу по убыванию значений в добавленном столбце `'debt_percentage'`.

In [29]:
children_debt_pivot['debt_percentage'] = children_debt_pivot[1]/(children_debt_pivot[0]+children_debt_pivot[1])*100
children_debt_pivot[1] = children_debt_pivot[1].astype(int)
children_debt_pivot[0] = children_debt_pivot[0].astype(int)
children_debt_pivot.sort_values(by='debt_percentage', ascending=False)

debt          0     1  debt_percentage
children                              
4            37     4         9.756098
2          1858   194         9.454191
1          4364   444         9.234609
3           303    27         8.181818
0         13028  1063         7.543822
5             9     0         0.000000

**Вывод:** Можно с осторожностью говорить о наличии зависимости между количеством детей и возвратом кредита в срок.
Наименьший процент просрочек (7.54%) наблюдается в группе бездетных клиентов. 
Среди клиентов, имеющих детей, наименьшая доля просрочек наблюдается в группе заемщиков, имеющих троих детей.
В группе заемщиков, имеющих 5 детей, не наблюдается ни одного случая просрочки возврата кредита. Поскольку данных по этой группе мало (всего 9 записей), полученные результаты могут быть нерепрезентативны. 

* *(дополнительные рассуждения: экспериментальные)* Следует отметить, что разница в 1-2% может иметь разный уровень значимости на разных объемах выборок. Попробуем применить критерий согласия хи-квадрат (либо точный критерий Фишера, если хотя бы одно из значений в таблице сопряженности менее 5) для сравнения распределения случаев возврата/невозврата кредита в срок для каджой группы по количеству детей с суммарными данными по остальным группам (порог значимости – 0.05). Добавим в таблицу `children_debt_pivot` столбец `'p_value'`, показывающий значение p для каждой группы по количеству детей, и столбец `'significance'`, показывающий наличие/отсутствия статистически значимого различия.
* Это можно сделать при помощи цикла или функций. Поэкспериментируем с обоими вариантами. Функции можно использовать и в других разделах

In [30]:
children_debt_0 = children_debt_pivot[0]
children_debt_1 = children_debt_pivot[1]
sum_debt_0 = sum(children_debt_0) # суммарные данные по случаям возврата кредита в срок
sum_debt_1 = sum(children_debt_1) # суммарные данные по случаям невозврата кредита в срок

children_p_value = []
children_significance = []

for i in range(len(children_debt_0)): 
    if (children_debt_0[i]<=5 or
        children_debt_1[i]<=5):
        oddsratio, pvalue = stats.fisher_exact([[children_debt_0[i], children_debt_1[i]], [sum_debt_0-children_debt_0[i], sum_debt_1-children_debt_1[i]]])
        children_p_value.append(pvalue)
        if pvalue<0.05:
            children_significance.append(1)
        else:
            children_significance.append(0)   
    else:
        obs = np.array([[children_debt_0[i], children_debt_1[i]], [sum_debt_0-children_debt_0[i], sum_debt_1-children_debt_1[i]]])
        chi_square, pvalue = stats.chi2_contingency(obs)[:2]
        children_p_value.append(pvalue)
        if pvalue<0.05:
            children_significance.append(1)
        else:
            children_significance.append(0)

children_debt_pivot['p_value'] = children_p_value
children_debt_pivot['significance'] = children_significance

children_debt_pivot

debt          0     1  debt_percentage   p_value  significance
children                                                      
0         13028  1063         7.543822  0.000020             1
1          4364   444         9.234609  0.001442             1
2          1858   194         9.454191  0.022276             1
3           303    27         8.181818  1.000000             0
4            37     4         9.756098  0.572455             0
5             9     0         0.000000  1.000000             0

In [31]:
def count_p_value(row): # функция для добавления столбца 'p_value'
    debt_0 = row[0]
    debt_1 = row[1]
    if debt_0<=5 or  debt_1<=5:
        oddsratio, pvalue = stats.fisher_exact([[debt_0, debt_1], [sum_debt_0-debt_0, sum_debt_1-debt_1]])
        return pvalue
    else:
        obs = np.array([[debt_0, debt_1], [sum_debt_0-debt_0, sum_debt_1-debt_1]])
        chi_square, pvalue = stats.chi2_contingency(obs)[:2]
        return pvalue

children_debt_pivot['p_value'] = children_debt_pivot.apply(count_p_value, axis=1)

def significance(row): # функция для добавления столбца 'significance'
    p_value = row['p_value']
    if p_value<0.05:
        return 1
    else:
        return 0

children_debt_pivot['significance'] = children_debt_pivot.apply(significance, axis=1)

children_debt_pivot

debt          0     1  debt_percentage   p_value  significance
children                                                      
0         13028  1063         7.543822  0.000020             1
1          4364   444         9.234609  0.001442             1
2          1858   194         9.454191  0.022276             1
3           303    27         8.181818  1.000000             0
4            37     4         9.756098  0.572455             0
5             9     0         0.000000  1.000000             0

* *Вывод дополнительный:* Статистически значимое различие в распределении случаев возврата/невозврата кредита в срок по отношению к суммарным данным по остальным групам наблюдается для групп заемщиков, не имеющих детей, и имеющих одного и двоих детей (Нулевую гипотезу можно  сформулировать следующим образом: *H0: Распределение случаев возврата/невозврата кредита в срок не связано с принадлежностью заемщика к определенной группе по количеству детей /не отличается существенно от "дефолтного" по всем группам заемщиков*). При этом, если учесть, что общая доля просрочек составляет около 8%, можно сделать вывод, что в группе бездетных заемщиков процент задолженностей по кредиту ниже, а в группе заемщиков с 1 и 2 детьми выше, чем в целом по выборке.

In [32]:
sum_debt_1/(sum_debt_0 + sum_debt_1)*100

8.119638085415593

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Построим сводную таблицу, показывающую соотношение случаев возврата и невозврата в срок кредита по группам клиентов с разным семейным положением. В данном разделе нам могут понадобиться данные столбцов `'family_status'` и `'family_status_id'`.
Cтолбец `'family_status'` содержит следующие значения: `'женат / замужем'`, `'гражданский брак'`, `'Не женат / не замужем'`,  `'в разводе'`, `'вдовец / вдова'`.
Cтолбец `'family_status_id'` содержит значения от 0 до 4.
Соответствия между значениями столбцов `'family_status'` и `'family_status_id'` можно установить, например, путем вызова метода `value_counts()` для этих столбцов: 

* 0 – `'женат / замужем'`
* 1 – `'гражданский брак'`,
* 2 – `'Не женат / не замужем'`, 
* 3 – `'в разводе'`, 
* 4 – `'вдовец / вдова'`.


Значения в столбце `'debt'`: 1 - наличие задолженности, 0 - отсутствие задолженности. 
В качестве значения аргумента `aggfunc` принимаем `'count'` - подсчитываем количество строк с каждым из значений в столбце `'debt'`.


In [33]:
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

In [34]:
data['family_status_id'].value_counts()

0    12261
1     4134
4     2796
3     1189
2      951
Name: family_status_id, dtype: int64

In [35]:
family_debt_pivot = data.pivot_table(index=['family_status_id', 'family_status'], columns='debt', values='education', aggfunc='count')
family_debt_pivot

debt                                        0    1
family_status_id family_status                    
0                женат / замужем        11334  927
1                гражданский брак        3749  385
2                вдовец / вдова           888   63
3                в разводе               1105   84
4                Не женат / не замужем   2523  273

In [36]:
sum(family_debt_pivot[0])+sum(family_debt_pivot[1])

21331

Добавим в таблицу `family_debt_pivot` столбец `'debt_percentage'`, показывающий долю долю случаев несвоевременного возврата кредита в каждой группе клиентов в соответствии с семейным положением, и отсортируем таблицу по убыванию значений в добавленном столбце `'debt_percentage'`.

In [37]:
family_debt_pivot['debt_percentage'] = family_debt_pivot[1]/(family_debt_pivot[0]+family_debt_pivot[1])*100
family_debt_pivot.sort_values(by='debt_percentage', ascending=False)

debt                                        0    1  debt_percentage
family_status_id family_status                                     
4                Не женат / не замужем   2523  273         9.763948
1                гражданский брак        3749  385         9.313014
0                женат / замужем        11334  927         7.560558
3                в разводе               1105   84         7.064760
2                вдовец / вдова           888   63         6.624606

**Вывод:** Процент несвоевременных выплат по кредиту выше в группах `'гражданский брак'` и `'Не женат / не замужем'`, в то время как заемщики с семейным статусом `'женат / замужем'`, `'в разводе'`, `'вдовец / вдова'` чаще возвращают кредиты в срок. Обобщая данные, можно предположить, что клиенты с опытом официально зарегистрированного брака реже имеют просрочки по кредитам, чем клиенты, не имеющие такого опыта.

* *(дополнительные рассуждения: экспериментальные)*

*Идея 1.* Здесь также можно попробовать применить критерий согласия хи-квадрат (*H0: Распределение случаев возврата/невозврата кредита в срок не связано с определенным семейным положением заемщика / для каждого семейного положения не отличается существенно от "дефолтного" по всем группам заемщиков*). Добавляем в таблицу `family_debt_pivot` столбцы, применяя уже готовые функции `count_p_value` и `significance`. 

На всякий случай проверим, что суммарные показатели совпадают с хранящимися в переменных `sum_debt_0` и `sum_debt_1`

In [38]:
print(sum_debt_0, sum_debt_1, sum_debt_0+sum_debt_1)
print(sum(family_debt_pivot[0]), sum(family_debt_pivot[1]), sum(family_debt_pivot[0])+sum(family_debt_pivot[1]))

19599 1732 21331
19599 1732 21331


In [39]:
family_debt_pivot['p_value'] = family_debt_pivot.apply(count_p_value, axis=1)
family_debt_pivot['significance'] = family_debt_pivot.apply(significance, axis=1)

family_debt_pivot

debt                                        0    1  debt_percentage   p_value  \
family_status_id family_status                                                  
0                женат / замужем        11334  927         7.560558  0.000560   
1                гражданский брак        3749  385         9.313014  0.001955   
2                вдовец / вдова           888   63         6.624606  0.095682   
3                в разводе               1105   84         7.064760  0.188231   
4                Не женат / не замужем   2523  273         9.763948  0.000731   

debt                                    significance  
family_status_id family_status                        
0                женат / замужем                   1  
1                гражданский брак                  1  
2                вдовец / вдова                    0  
3                в разводе                         0  
4                Не женат / не замужем             1

*Вывод дополнительный 1:* Можно констатировать наличие статистически значимого различия в распределении случаев возврата/невозврата кредита в срок для каждой из групп заемщиков со статусом `'женат / замужем'`, `'гражданский брак'`, `'Не женат / не замужем'` по отношению к суммарным данным по остальным групам. Для групп `'в разводе'`, `'вдовец / вдова'` статистически значимого различия не наблюдается. Возможно, это объясняется тем, что они меньше по объему.

*Идея 2.* Выше было высказано предположение о существовании зависимости между опытом официальной регистрации брака и возвратом кредита в срок. Можно попробовать объединить данные в две группы в зависимости от того, регистрировали ли заемщики брак (категория 'A') или нет (категория 'В'). Построим сводную таблицу, которая будет иметь вид таблицы сопряженности 2 * 2 и применим к значениям в ней критерий согласия хи-квадрат (выведем только значения хи-квадрит и p). 

In [40]:
def categirize_family_status(row):
    if 'женат / замужем' in row or 'вдовец / вдова' in row or 'в разводе' in row:
        return 'A'
    elif 'Не женат / не замужем' in row or 'гражданский брак' in row:
        return 'B'

data['family_status_category'] = data['family_status'].apply(categirize_family_status)
family_category_debt_pivot = data.pivot_table(index=['family_status_category'], columns='debt', values='education', aggfunc='count')
family_category_debt_pivot['debt_percentage'] = family_category_debt_pivot[1]/(family_category_debt_pivot[0]+family_category_debt_pivot[1])*100

display(family_category_debt_pivot)
stats.chi2_contingency([[family_category_debt_pivot.loc['A', 0], family_category_debt_pivot.loc['A', 1]], 
                        [family_category_debt_pivot.loc['B', 0], family_category_debt_pivot.loc['B', 1]]])[:2]

debt                        0     1  debt_percentage
family_status_category                              
A                       13327  1074         7.457815
B                        6272   658         9.494949

(25.752881674877585, 3.880471871323918e-07)

*Вывод дополнительный 2:* Заемщики, которые заключали официальный брак, чаще возвращают кредит в срок, чем заемщики, которые не оформляли отношения. Это различие статистически значимо.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Построим сводную таблицу, показывающую соотношение случаев возврата и невозврата в срок кредита по группам клиентов с разным уровнем дохода. В данном разделе нам потребуются данные столбца `'total_income_category'`. Напомним, что уровень дохода в данном столбце определяется при помощи букв A-E.
Значения в столбце `'debt'`: 1 - наличие задолженности, 0 - отсутствие задолженности. 
В качестве значения аргумента `aggfunc` принимаем `'count'` - подсчитываем количество строк с каждым из значений в столбце `'debt'`.

In [41]:
income_debt_pivot = data.pivot_table(index=['total_income_category'], columns='debt', 
                                     values='education', aggfunc='count')
income_debt_pivot

debt                       0     1
total_income_category             
A                         23     2
B                       4660   354
C                      14568  1353
D                        328    21
E                         20     2

Добавим в таблицу `income_debt_pivot` столбец `'debt_percentage'`, показывающий долю долю случаев несвоевременного возврата кредита в каждой группе клиентов в соответствии с семейным положением, и отсортирует таблицу по убыванию значений в добавленном столбце `'debt_percentage'`.

In [42]:
income_debt_pivot['debt_percentage'] = income_debt_pivot[1]/(income_debt_pivot[0]+income_debt_pivot[1])*100
income_debt_pivot.sort_values(by='debt_percentage', ascending=False)

debt                       0     1  debt_percentage
total_income_category                              
E                         20     2         9.090909
C                      14568  1353         8.498210
A                         23     2         8.000000
B                       4660   354         7.060231
D                        328    21         6.017192

**Вывод:** Наименьшее количество заемщиков представлено в группах с наиболее высоким (категория `'А'`) и наиболее низким (категория `'Е'`) уровнем дохода. На основании таких небольших выборок выводы следует делать с осторожностью, результаты могут быть недостаточно репрезентативны.
В целом, сложно сделать однозначный вывод о наличии зависимости между уровнем дохода и возвратом кредита в срок. 
Наименьший процент несвоевременных выплат отмечается в категории `'D'` (средний уровень дохода, 30001–50000). 
Наиболее многочисленными являются группы заемщиков с доходом выше среднего (категория `'С'`, 50001–200000) и высоким уровнем дохода (категория `'В'`, 200001–1000000). При этом процент просрочек выше в категории `'С'` (уровень дохода выше среднего, 50001–200000), т.е. можно сделать вывод, что клиенты с высоким доходом чаще отдают кредит в срок, чем клиенты с дохоом выше среднего.

* *(дополнительные рассуждения: экспериментальные)* 
Здесь также можно попробовать применить критерий согласия хи-квадрат (*H0: Распределение случаев возврата/невозврата кредита в срок не связано с категорией дохода заемщика / для каждой категории не отличается существенно от "дефолтного" по всем группам заемщиков*). Добавляем в таблицу `income_debt_pivot` столбцы, применяя уже готовые функции `count_p_value` и `significance`.

In [43]:
income_debt_pivot['p_value'] = income_debt_pivot.apply(count_p_value, axis=1)
income_debt_pivot['significance'] = income_debt_pivot.apply(significance, axis=1)

income_debt_pivot

debt                       0     1  debt_percentage   p_value  significance
total_income_category                                                      
A                         23     2         8.000000  1.000000             0
B                       4660   354         7.060231  0.001867             1
C                      14568  1353         8.498210  0.000574             1
D                        328    21         6.017192  0.176661             0
E                         20     2         9.090909  0.698223             0

*Вывод дополнительный 2:* о существенных различиях в распределении случаев возврата/невозврата кредита в срок можно говорить только применительно к группам `'B'` - высокий доход и `'С'` - доход выше среднего. При высоком доходе доля случаев несвоевременного выполнения кредитных обязательств ниже.

### Как разные цели кредита влияют на его возврат в срок?

Построим сводную таблицу, показывающую соотношение случаев возврата и невозврата в срок кредита в зависимости от цели кредита. В данном разделе нам потребуются данные столбца `'purpose_category'`. Напомним, что в столбце `'purpose_category'` имеются следующие категории: `'операции с автомобилем'`, `'операции с недвижимостью'`, `'проведение свадьбы'`, `'получение образования'`.
Значения в столбце `'debt'`: 1 - наличие задолженности, 0 - отсутствие задолженности. 
В качестве значения аргумента `aggfunc` принимаем `'count'` - подсчитываем количество строк с каждым из значений в столбце `'debt'`.

In [44]:
purpose_debt_pivot = data.pivot_table(index=['purpose_category'], columns='debt', 
                                      values='education', aggfunc='count')
purpose_debt_pivot

debt                         0    1
purpose_category                   
операции с автомобилем    3879  400
операции с недвижимостью  9971  780
получение образования     3619  369
проведение свадьбы        2130  183

Добавим в таблицу `purpose_debt_pivot` столбец `'debt_percentage'`, показывающий долю долю случаев несвоевременного возврата кредита в каждой группе клиентов в соответствии с семейным положением, и отсортирует таблицу по убыванию значений в добавленном столбце `'debt_percentage'`.

In [45]:
purpose_debt_pivot['debt_percentage'] = purpose_debt_pivot[1]/(purpose_debt_pivot[0]+purpose_debt_pivot[1])*100
purpose_debt_pivot.sort_values(by='debt_percentage', ascending=False)

debt                         0    1  debt_percentage
purpose_category                                    
операции с автомобилем    3879  400         9.347978
получение образования     3619  369         9.252758
проведение свадьбы        2130  183         7.911803
операции с недвижимостью  9971  780         7.255139

**Вывод:** Наименьшая доля несвоевременных выплат по кредиту наблюдается в тех случаях, когда кредит берется на операции с недвижимостью. Возможное объяснение: как правило, кредит на операции с недвижимостью берется на долгий срок, а также требует начичия у клиента определенного кочичества собственных средств. В таких случаях заемщики более тщательно оценивают свои финансовые возможности при перед получением кредита и реже допускают просрочки.
Более высокий процент просрочек наблюдается при совершении операций с автомобилем и при получении кредита на образование.

* *(дополнительные рассуждения: экспериментальные)* 
Здесь также можно попробовать применить критерий согласия хи-квадрат. Добавляем в таблицу `purpose_debt_pivot` столбцы, применяя уже готовые функции `count_p_value` и `significance`.

In [46]:
purpose_debt_pivot['p_value'] = purpose_debt_pivot.apply(count_p_value, axis=1)
purpose_debt_pivot['significance'] = purpose_debt_pivot.apply(significance, axis=1)

purpose_debt_pivot

debt                         0    1  debt_percentage   p_value  significance
purpose_category                                                            
операции с автомобилем    3879  400         9.347978  0.001118             1
операции с недвижимостью  9971  780         7.255139  0.000004             1
получение образования     3619  369         9.252758  0.004062             1
проведение свадьбы        2130  183         7.911803  0.728397             0

*Вывод дополнительный:* подтверждается гипотеза о большей вероятности возврата кредита в долг в случаях, если цель кредита – операции с жильем, и меньшей – если кредит взят на покупку автомобиля или получение образования.

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Изначально пропуски присутствовали в двух столбцах: `days_employed` и `total_income`. Интересно, что пропусков в обоих случаях было одинаковое количество – 2174. Возможно, при отсутствии информации в одном столбце, второй автоматически оставался незаполненным (например, учитывается доход только при наличии официального трудоустройства/стажа). Загрузим исходный  датесет еще раз и попробуем определить, являются ли пропуски случайными. 

In [47]:
try:
    data_raw = pd.read_csv('/datasets/data.csv')
except:
    data_raw = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')


Гипотеза о том, что пропуски в двух столбцах каким-то образом связаны, в целом подтверждаются (см. ниже вывод первых 20 строк с пропусками). При этом  причины появления пропусков неочевидны: это клиенты различного пола и возраста, с разным уровнем образования, семейным положением, типом дохода, целью получения кредита и т.д. О связи с официальным трудоустройством также сложно говорить (например, в выборке есть госслужащие).

In [48]:
data_na = data_raw[data_raw['days_employed'].isna()]
data_na.head(20)

children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
65          0            NaN         21   среднее             1   
67          0            NaN         52    высшее             0   
72          1            NaN         32    высшее             0   
82          2            NaN         50    высшее             0   
83          0            NaN         52   среднее             1   
90          2            NaN         35    высшее             0   
94          1            NaN         34    высшее             0   
96          0            NaN         44   СРЕДНЕЕ             1   
97          0            NaN         47    высшее             0   
120         0            NaN         46   среднее             1   
121         0            NaN         29    высшее             0   
135         0            NaN         27   среднее             1   
141         0            NaN         39   среднее             1   
145         0            NaN         62   среднее             1   
174         0            NaN         55    высшее             0   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
65   Не женат / не замужем                 4      M    компаньон     0   
67         женат / замужем                 0      F    пенсионер     0   
72         женат / замужем                 0      M  госслужащий     0   
82         женат / замужем                 0      F    сотрудник     0   
83         женат / замужем                 0      M    сотрудник     0   
90         женат / замужем                 0      F    сотрудник     0   
94        гражданский брак                 1      F    компаньон     0   
96         женат / замужем                 0      F    сотрудник     0   
97         женат / замужем                 0      F    сотрудник     0   
120        женат / замужем                 0      F    сотрудник     0   
121        женат / замужем                 0      F    компаньон     0   
135        женат / замужем                 0      M    компаньон     0   
141       гражданский брак                 1      M    сотрудник     0   
145        женат / замужем                 0      M    пенсионер     0   
174         вдовец / вдова                 2      F    компаньон     0   

     total_income                                purpose  
12            NaN                        сыграть свадьбу  
26            NaN                            образование  
29            NaN       строительство жилой недвижимости  
41            NaN       сделка с подержанным автомобилем  
55            NaN                        сыграть свадьбу  
65            NaN  операции с коммерческой недвижимостью  
67            NaN                покупка жилья для семьи  
72            NaN  операции с коммерческой недвижимостью  
82            NaN                                  жилье  
83            NaN                                  жилье  
90            NaN                      операции с жильем  
94            NaN                  на проведение свадьбы  
96            NaN             покупка жилой недвижимости  
97            NaN                 профильное образование  
120           NaN                     высшее образование  
121           NaN                             автомобиль  
135           NaN                      

In [49]:
data_na['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* медианное значение более корректно показывает среднюю величину при большом разбросе количественных показателей в выборке и при наличии значительных выбросов. В случае с количеством дней стажа и уровнем дохода среднее арифметическое и медиана могут существенно различаться.

Приведем максимальные и минимальные значения в столбцах `days_employed` и `total_income`. Поскольку в задании 13 тип данных в `total_income` был заменен на целочисленный, значения из данного столбца выводятся в виде целого числа. 

Минимальное значение в столбце `total_income` на два порядка меньше, чем максимальное, что подтверждает большой разброс значений. Кроме того, количество заемщиков с низким (E) и очень высоким доходом (А) невелико (см. раздел 3.2).

В столбце `days_employed` как минимальное, так и максимальное значение выглядят неправдоподобно. Использование среднего арифметического в качестве средней величины в таком случае может привести к искаженному представлению о данных.

In [50]:
print('total_income', 'min =', data['total_income'].min(), 'max =', data['total_income'].max())
print('days_employed', 'min =', data['days_employed'].min(), 'max =', data['days_employed'].max())

total_income min = 20667 max = 2265604
days_employed min = 24.14163324048118 max = 401755.40047533


## Общий вывод

На основе статистики о платёжеспособности клиентов была исследована возможное влияние различных факторов на факт возврата кредита в срок.

В ходе проекта были реализованы следующие **этапы**:
1. данные загружены и изучены;
2. проведена предобработка данных (удаление пропусков, обработка аномальных значений, изменение типов данных, обработка дубликатов, категоризация данных);
3. осуществлен анализ зависимости между фактом возврата кредита в срок и такими параметрами как количество детей, семейное положение, уровень дохода, цель кредита.


**Анализ показывает**, что для всех рассмотренных признаков данных наблюдается зависимость между принадлежностью заемщиков к определенным группам возвратом кредита в срок:
* *в зависимости от количества детей*: бездетные заемщики чаще отдают кредит в срок, чем клиенты с 1-2 детьми;
* *в зависимости от семейного положения*: заемщики, состоящие (или ранее состоявшие) в браке чаще отдают кредит в срок, чем клиенты, не регистрировавшие отношения;
* *в зависимости от уровня дохода*: заемщики с высоким уровнем дохода чаще возвращают кредит в срок
* *в зависимости от целей кредита*: заемщики, берущие кредит на опрерации с недвижимостью, чаще возвращают кредит в срок, чем клиенты, берущие кредит на другие цели

Таким образом, идеальным заемщиком является клиент банка, состоящий в официальном браке, но не имеющий детей, имеющий высокий уровень дохода и берущий кредит с целью приобретения недвижимости.